aten/src/ATen/core/dispatch/Dispatcher.h

```diff
@@ -429,20 +432,24 @@ inline void Dispatcher::callBoxed(const OperatorHandle& op, Stack* stack) const
   const auto& kernel = entry.lookup(dispatchKey);
 
 #ifndef PYTORCH_DISABLE_PER_OP_PROFILING
-  // using already existing stack to record function execution in observers
-  at::RecordFunction guard(at::RecordScope::FUNCTION);
-  if (C10_UNLIKELY(guard.isActive())) {
-    if (shouldRecord(dispatchKey) && entry.isObserved()) {
-      int64_t seq_num = -1;
-      if (isIncludedInAlias(dispatchKey, DispatchKey::Autograd) && at::GradMode::is_enabled()) {
-        seq_num = at::sequence_number::peek();
-      }
-      if (guard.needs_inputs) {
-        guard.before(op, *stack, seq_num);
-      } else {
-        guard.before(op, seq_num);
+  if (at::shouldRunRecordFunction()) {
```

Comment:
is performance of the case where we *do* need to record all functions important? This should be a regression in that case because of (sadly) the extra function call overhead.

Reply:
it's important for specific use cases (mainly, profiler), that use this; but normally in production profiler won't be enabled; existing production observers are already heavy sampled; in this pr we move sampling partially up, into Dispatcher.h, to avoid creating RecordFunction object altogether